specifyweb/specify/scoping.py

```diff
@@ -1,96 +1,134 @@
+        if has_related(self.obj, 'institution'):
+            return ScopeType.INSTITUTION, self.obj.institution
 
     def taxon(self):
-        return DISCIPLINE_SCOPE, self.obj.definition.discipline.id
+        return ScopeType.DISCIPLINE, self.obj.definition.discipline
 
 #############################################################################
 
-    def _simple_discipline_scope(self):
-        return DISCIPLINE_SCOPE, self.obj.discipline_id
-
-    def _simple_division_scope(self):
-        return DIVISION_SCOPE, self.obj.division_id
+    def _simple_discipline_scope(self) -> Tuple[int, Model]:
+        return ScopeType.DISCIPLINE, self.obj.discipline
+
+    def _simple_division_scope(self) -> Tuple[int, Model]:
+        return ScopeType.DIVISION, self.obj.division
+
+    def _simple_collection_scope(self) -> Tuple[int, Model]:
+        if hasattr(self.obj, "collectionmemberid"):
+            try:
+                """
+                    Collectionmemberid is not a primary key, but a plain 
+                    numerical field, meaning the Collection it is 
+                    supposed to 'reference' may not exist anymore
+                """
+                collection = models.Collection.objects.get(
+                    pk=self.obj.collectionmemberid)
+            except ObjectDoesNotExist:
+                if not hasattr(self.obj, "collection"):
+                    raise
+                collection = self.obj.collection
+        else:
+            collection = self.obj.collection
 
-    def _simple_collection_scope(self):
-        return COLLECTION_SCOPE, self.obj.collectionmemberid
+        return ScopeType.COLLECTION, collection
 
     def _infer_scope(self):
-        if hasattr(self.obj, "division_id"): return self._simple_division_scope()
-        if hasattr(self.obj, "discipline_id") : return self._simple_discipline_scope()
-        if hasattr(self.obj, "collectionmemberid"): return self._simple_collection_scope()
+        if has_related(self.obj, "division"):
+            return self._simple_division_scope()
+        if has_related(self.obj, "discipline"):
+            return self._simple_discipline_scope()
+        if has_related(self.obj, "collectionmemberid") or has_related(self.obj, "collection"):
+            return self._simple_collection_scope()
+
+        return self._default_institution_scope()
 
     # If the table has no scope, and scope can not be inferred then scope to institution
-    def _default_institution_scope(self):
+    def _default_institution_scope(self) -> Tuple[int, Model]:
         institution = models.Institution.objects.get()
-        return INSTITUTION_SCOPE, institution.id
+        return ScopeType.INSTITUTION, institution
+
+
+def has_related(model_instance, field_name: str) -> bool:
+    return hasattr(model_instance, field_name) and getattr(model_instance, field_name, None) is not None
+
+
+def in_same_scope(object1: Model, object2: Model) -> bool:
+    """
```

Comment:
same as above

Reply:
Fixed in https://github.com/specify/specify7/pull/4215/commits/43aee5b97e52edca55b58e0cb2b76773754b510f